File: source/python/pipeline.py
import threading
import queue

gen_q = queue.Queue(maxsize=5)
dbl_q = queue.Queue(maxsize=5)
add_q = queue.Queue(maxsize=5)

def generate(out):
    for i in range(1, 6):
        out.put(i)
    out.put(None)

def double_stage(inp, out):
    while True:
        val = inp.get()
        if val is None:
            out.put(None)
            break
        out.put(val * 2)

def add10_stage(inp, out):
    while True:
        val = inp.get()
        if val is None:
            out.put(None)
            break
        out.put(val + 10)

t1 = threading.Thread(target=generate, args=(gen_q,))
t2 = threading.Thread(target=double_stage, args=(gen_q, dbl_q))
t3 = threading.Thread(target=add10_stage, args=(dbl_q, add_q))

t1.start()
t2.start()
t3.start()

results = []
while True:
    val = add_q.get()
    if val is None:
        break
    results.append(val)

t1.join()
t2.join()
t3.join()

print(" ".join(str(x) for x in results))

File: source/go/pipeline.go
package main

import (
	"fmt"
	"strings"
)

func generate(out chan<- int) {
	for i := 1; i <= 5; i++ {
		out <- i
	}
	close(out)
}

func doubleStage(input <-chan int, output chan<- int) {
	for val := range input {
		output <- val * 2
	}
	close(output)
}

func add10Stage(input <-chan int, output chan<- int) {
	for val := range input {
		output <- val + 10
	}
	close(output)
}

func main() {
	genCh := make(chan int, 5)
	dblCh := make(chan int, 5)
	addCh := make(chan int, 5)

	go generate(genCh)
	go doubleStage(genCh, dblCh)
	go add10Stage(dblCh, addCh)

	var results []string
	for val := range addCh {
		results = append(results, fmt.Sprintf("%d", val))
	}

	fmt.Println(strings.Join(results, " "))
}
